### ISSUE TITLE
API Gateway- Websocket

### ISSUE THREAD

**Issue by bbernays** [terraform-providers/terraform-provider-aws#7004]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

API Gateway now supports WebSocket connections. It would be great to be able to define this resource 

### New or Affected Resource(s)

* aws_api_gateway_v2
* aws_api_gateway_v2_route
* aws_api_gateway_v2_integration
* aws_api_gateway_v2_integration_response
* aws_api_gateway_v2_route_response
* aws_api_gateway_v2_deployment
* aws_api_gateway_v2_stage

### Potential Terraform Configuration

```hcl
#Incomplete!!
resource "aws_api_gateway_v2" "websocket" {
  name = "myWebSocketApi3"
  description    = "${format("CMS V2 %s", var.environment)}"
  protocol_type = "WEBSOCKET"
  route_selection_expression = "$request.body.action'"
}
resource "aws_api_gateway_v2_route" "non_default_route" {
  api_id = "${aws_api_gateway_v2.websocket.id}"
  routeKey = "testing"
}


```

### References

* [Documentation](https://docs.aws.amazon.com/apigateway/latest/developerguide/apigateway-websocket-api.html)

* [Announcement](https://aws.amazon.com/blogs/compute/announcing-websocket-apis-in-amazon-api-gateway/)

**Comment by mbonig** [comment]
I'd like to contribute to this, but have never done any development on this project before. Can we explain a little further the changes that are needed?

Right now I read this to mean there are new _v2 resources created, and that they will largely be similar to the existing resources but with additional logic. But, I think just adding support to the existing resources, with intelligent defaults, feels like a better fit. 

Either way, I'll try submitting a PR...

**Comment by bbernays** [comment]
I suggested that it be _v2 to stay inline with AWS naming of the new resource. [Link here](https://docs.aws.amazon.com/apigatewayv2/latest/api-reference/api-reference.html)

**Comment by mbonig** [comment]
I've created a fork and made some changes to support the first new 'v2' resource. Anyone care to review and let me know things I got wrong before implementing the rest?

https://github.com/mbonig/terraform-provider-aws/tree/f-gatewayv2

Thanks!

**Comment by mbonig** [comment]
Additional affected resources:
* aws_api_gateway_v2_authorizer

**Comment by krotscheck** [comment]
Also affected: aws_api_gateway_domain_name - may need a new resource, I'm not familiar with the underlying api.

**Comment by mbonig** [comment]
I'm working on this, but on the _route_ object I've come across an issue. I want the terraform to look like this:
```
resource "aws_api_gateway_v2_route" "test_route" {
   ...
   request_parameters {
       "asdf.asdf.asdf.asdf" {
          required=true
       }
   }
}
```

as this seems to be the best representation of the underlying data. However, I can't seem to find a schema definition that will work with this. Does anybody have any tips, or can point me to an existing resource that has a similar setup? 

API reference: 
https://docs.aws.amazon.com/sdk-for-go/api/service/apigatewayv2/#CreateRouteInput

**Comment by Jumblemuddle** [comment]
Has there been any update on this? I'm willing to help with testing PRs.

**Comment by mbonig** [comment]
I have had priorities shift so I haven't been able to work on it. Would be willing to give someone my branch if they want to continue my work...

**Comment by sidewaiise** [comment]
I'd be happy to contribute. What's the turnaround time for PR / Testing / Approval? I need this feature actually in the next couple of weeks.

**Comment by c1moore** [comment]
Has there been any progress on this issue?  It looks like it's the second most popular issue.  @mbonig could you post your work so somebody else can pick up from it?

**Comment by ewbankkit** [comment]
I have made a start on this, based on the initial work from @mbonig. I think a separate PR for each new resource is best:

* [`aws_api_gateway_v2_api`](https://github.com/terraform-providers/terraform-provider-aws/pull/8842), needs to be merged before subsequent resources
* `aws_api_gateway_v2_api_mapping`
* `aws_api_gateway_v2_authorizer`
* `aws_api_gateway_v2_deployment`
* `aws_api_gateway_v2_domain_name`
* `aws_api_gateway_v2_integration`
* `aws_api_gateway_v2_integration_response`
* `aws_api_gateway_v2_model`
* `aws_api_gateway_v2_route`
* `aws_api_gateway_v2_route_response`
* `aws_api_gateway_v2_stage`

**Comment by rpstreef** [comment]
Is there an ETA when this will be released?

**Comment by ewbankkit** [comment]
I think we need to decide on the set of resources that comprise an "MVP" and focus on those - @rpstreef, what would be the minimal set of resources for your use case(s)? Thanks.

**Comment by rpstreef** [comment]
I wont be using;
aws_api_gateway_v2_api_mapping
aws_api_gateway_v2_model
aws_api_gateway_v2_domain_name

When you use proxy integration, these are not needed
aws_api_gateway_v2_integration_response
aws_api_gateway_v2_route_response

Everything else is MVP in my opinion

**Comment by ewbankkit** [comment]
API Gateway v2 tagging added in [AWS SDK v.20.9](https://github.com/aws/aws-sdk-go/releases/tag/v1.20.9).
Requires:
* https://github.com/terraform-providers/terraform-provider-aws/pull/9153

**Comment by ewbankkit** [comment]
Tags are [supported](https://docs.aws.amazon.com/apigateway/latest/developerguide/apigateway-tagging-supported-resources.html) on:
* `aws_api_gateway_v2_api`
* `aws_api_gateway_v2_domain_name`
* `aws_api_gateway_v2_stage`

**Comment by rpstreef** [comment]
> 
> 
> Is there an ETA when this will be released?

for the MVP, do we have an ETA?

**Comment by ewbankkit** [comment]
OK, I think I have the resources for @rpstreef's MVP done.
The PR will need to be reviewed and merged in the following order with PRs in later phases rebased:

#### Phase 1

* `aws_api_gateway_v2_api` - https://github.com/terraform-providers/terraform-provider-aws/pull/8842

#### Phase 2

* `aws_api_gateway_v2_authorizer` - https://github.com/terraform-providers/terraform-provider-aws/pull/9228
* `aws_api_gateway_v2_integration` - https://github.com/terraform-providers/terraform-provider-aws/pull/8949
* `aws_api_gateway_v2_model` - https://github.com/terraform-providers/terraform-provider-aws/pull/8912

#### Phase 3

* `aws_api_gateway_v2_route` - https://github.com/terraform-providers/terraform-provider-aws/pull/8881

#### Phase 4

* `aws_api_gateway_v2_deployment` - https://github.com/terraform-providers/terraform-provider-aws/pull/9245

#### Phase 5

* `aws_api_gateway_v2_stage` - https://github.com/terraform-providers/terraform-provider-aws/pull/9232

**Comment by nywilken** [comment]
hi @ewbankkit thanks for pushing forward the pull-requests for these resources. We don't have an ETA on when we can start to review or merge in these changes but happy to update the thread once we have a better idea. In the meantime, it would be helpful if there was an outline on the immediate problem the said resources would address along with one or more expected use cases so that the maintainers can get a better sense of how we should proceed when starting our review. 

Cheers!

**Comment by rpstreef** [comment]
@nywilken This is the top 4 most voted issue and by date (December) also the longest waiting for a resolution, 8 months by now. According to your own prioritization rules this should get priority to be reviewed and released.
Now I'm not trying to be smart or insensitive to your workloads, but the community is waiting on this to get done fairly soon.

**Comment by ewbankkit** [comment]
My current thinking is to
* Complete the outstanding resources, `aws_api_gateway_v2_domain_name` and ` `aws_api_gateway_v2_api_mapping`
* Add an example to https://github.com/terraform-providers/terraform-provider-aws/tree/master/examples that duplicates https://github.com/aws-samples/simple-websockets-chat-app and use that as the use case for the initial set of resources to review
* Add a more fully featured example that might use all the new resources

**Comment by nywilken** [comment]
@rpstreef these are all valid points, and rest assure we are aware of the votes. Please don't take my request for information as a deprioritization step. Information on what problem(s) the new resource solves and, more importantly, use cases on how users expect to interact with it via Terraform helps maintainers understand the implementation approach during the time of review.

For example, use cases such as expecting to work with continuous deployment tooling after initial Terraform deployment would imply that we may want to look at prioritizing data sources or ensuring resources have certain attributes for the external systems to play nice. In the case of dealing with Lambda services we may need to take a deep look at how things are versioned, if any.

Newly submitted resources without this context requires additional thinking around the proposed design, which we have to extract and piece together through the code. While the latter is possible it does add more time to the review. So it is always best to have more information upfront to prepare maintainers in supporting and advising on the contribution.

**Comment by john-osullivan** [comment]
@nywilken  Just found this issue when I was looking for docs on the new Websocket API Gateway' Terraform resource.  My intended usecase is an endpoint which lets me open multiple websocket connections, some of them roughly in a master-slave configuration.

Let's say that Alice connected to `websocket-api-endpoint.com/v1/`, they would be redirected a random route like `.../v1/human-friendly-name` and would get a live session with our application.  If Bob then went to that same `user-friendly-name` route while Alice's session is open, then they would see Alice's current state of using our application.  Kind of like dynamically creating chat rooms, except that Alice, the "creator", is doing most of the message sending.  Bob will occasionally want to be able to send messages that get to Alice, though.

Our compute infra is strictly Lambda functions as much as we can manage it, so having that out of the box would be huge!  We may eventually want to be able to update the Lambda function in-place with a CI/CD pipeline, but for now we've been alright running a Terraform apply with fresh source.

Glad to see there's progress on getting this supported, watching this issue.

**Comment by tqhoughton** [comment]
What is the status on this? I would like to be able to create websocket apis with terraform, can you give us an ETA on when this will be merged and available in a release?

**Comment by ewbankkit** [comment]
With the [API Gateway WebSocket chat app example](https://github.com/terraform-providers/terraform-provider-aws/pull/9533) as a simple end-to-end working example, I suggest focusing on the following resources initially:
1.  `aws_api_gateway_v2_api` - https://github.com/terraform-providers/terraform-provider-aws/pull/8842
1. `aws_api_gateway_v2_integration` - https://github.com/terraform-providers/terraform-provider-aws/pull/8949
1. `aws_api_gateway_v2_route` - https://github.com/terraform-providers/terraform-provider-aws/pull/8881. I will need to comment out the acceptance tests that use `aws_api_gateway_v2_authorizer` - https://github.com/terraform-providers/terraform-provider-aws/pull/9228 and `aws_api_gateway_v2_model` - https://github.com/terraform-providers/terraform-provider-aws/pull/8912
1. `aws_api_gateway_v2_deployment` - https://github.com/terraform-providers/terraform-provider-aws/pull/9245
1.  `aws_api_gateway_v2_stage` - https://github.com/terraform-providers/terraform-provider-aws/pull/9232

**Comment by rpstreef** [comment]
We're almost 10 months in for this issue, when can we expect a first release?

**Comment by tqhoughton** [comment]
Yeah I ended up going with a cloudformation template, which is not ideal but it's better than trying to string a bunch of CLI scripts together. :/

This is a template that deploys a websocket API with an authorizer:

```
resource "aws_cloudformation_stack" "websocket_api" {
  name = "${var.websocket_api_stack_name}"

  depends_on = [
    "module.websockets_service_onconnect_lambda",
    "module.websockets_service_ondisconnect_lambda",
    "module.websockets_service_authorizer_lambda"
  ]

  template_body = <<EOF
Resources:
  WebSocketApi:
    Type: AWS::ApiGatewayV2::Api
    Properties:
      Name: ${var.websocket_api_stack_name}
      ProtocolType: WEBSOCKET
      RouteSelectionExpression: $request.body.action
  
  WebSocketAuth:
    Type: "AWS::ApiGatewayV2::Authorizer"
    Properties:
      Name: lambda-auth
      ApiId: !Ref WebSocketApi
      AuthorizerType: REQUEST
      AuthorizerUri: arn:aws:apigateway:${var.target_aws_region}:lambda:path/2015-03-31/functions/${module.websockets_service_authorizer_lambda.function_arn}/invocations
      IdentitySource:
        - "route.request.querystring.Authorization"
  
  AuthorizerFunctionPermission:
    Type: AWS::Lambda::Permission
    DependsOn: [ WebSocketApi ]
    Properties:
      Action: lambda:InvokeFunction
      FunctionName: ${module.websockets_service_authorizer_lambda.function_name}
      Principal: apigateway.amazonaws.com
  
  OnConnectRouteIntegration:
    Type: AWS::ApiGatewayV2::Integration
    Properties:
      ApiId: !Ref WebSocketApi
      IntegrationType: AWS_PROXY
      IntegrationUri: arn:aws:apigateway:${var.target_aws_region}:lambda:path/2015-03-31/functions/${module.websockets_service_onconnect_lambda.function_arn}/invocations

  OnConnectApiRoute:
    Type: AWS::ApiGatewayV2::Route
    Properties:
      ApiId: !Ref WebSocketApi
      RouteKey: $connect
      AuthorizationType: CUSTOM
      AuthorizerId: !Ref WebSocketAuth
      OperationName: connect
      Target: !Join ['/', [ integrations, !Ref OnConnectRouteIntegration ] ]
  
  OnConnectFunctionPermission:
    Type: AWS::Lambda::Permission
    DependsOn: [ WebSocketApi ]
    Properties:
      Action: lambda:invokeFunction
      FunctionName: ${module.websockets_service_onconnect_lambda.function_name}
      Principal: apigateway.amazonaws.com
  
  OnDisconnectRouteIntegration:
    Type: AWS::ApiGatewayV2::Integration
    Properties:
      ApiId: !Ref WebSocketApi
      IntegrationType: AWS_PROXY
      IntegrationUri: arn:aws:apigateway:${var.target_aws_region}:lambda:path/2015-03-31/functions/${module.websockets_service_ondisconnect_lambda.function_arn}/invocations

  OnDisconnectApiRoute:
    Type: AWS::ApiGatewayV2::Route
    Properties:
      ApiId: !Ref WebSocketApi
      RouteKey: $disconnect
      AuthorizationType: NONE
      OperationName: disconnect
      Target: !Join ['/', [ integrations, !Ref OnDisconnectRouteIntegration ] ]

  OnDisconnectFunctionPermission:
    Type: AWS::Lambda::Permission
    DependsOn: [ WebSocketApi ]
    Properties:
      Action: lambda:invokeFunction
      FunctionName: ${module.websockets_service_ondisconnect_lambda.function_name}
      Principal: apigateway.amazonaws.com
  
  DefaultRouteIntegration:
    Type: AWS::ApiGatewayV2::Integration
    Properties:
      ApiId: !Ref WebSocketApi
      IntegrationType: AWS_PROXY
      IntegrationUri: arn:aws:apigateway:${var.target_aws_region}:lambda:path/2015-03-31/functions/${module.websockets_service_default_lambda.function_arn}/invocations

  DefaultApiRoute:
    Type: AWS::ApiGatewayV2::Route
    Properties:
      ApiId: !Ref WebSocketApi
      RouteKey: $default
      AuthorizationType: NONE
      OperationName: default
      Target: !Join ['/', [ integrations, !Ref DefaultRouteIntegration ] ]

  DefaultFunctionPermission:
    Type: AWS::Lambda::Permission
    DependsOn: [ WebSocketApi ]
    Properties:
      Action: lambda:invokeFunction
      FunctionName: ${module.websockets_service_default_lambda.function_name}
      Principal: apigateway.amazonaws.com
Outputs:
  WebSocketApi:
    Value: !Ref WebSocketApi
EOF
}

resource "null_resource" "create_stage" {
  depends_on = [
    "aws_cloudformation_stack.websocket_api"
  ]

  triggers {
    # DO NOT CHANGE THIS
    invocation = "2019-08-26:T10:00:00Z"
  }

  provisioner "local-exec" {
    command = "aws apigatewayv2 create-stage --api-id ${aws_cloudformation_stack.websocket_api.outputs["WebSocketApi"]} --stage-name ${var.websocket_api_stage}"
  }
}

resource "null_resource" "create_deployment" {
  depends_on = [
    "null_resource.create_stage"
  ]

  triggers {
    lambda = "${md5(file("./terraform/apigateway_v2.tf"))}"
  }

  provisioner "local-exec" {
    command = "aws apigatewayv2 create-deployment --api-id ${aws_cloudformation_stack.websocket_api.outputs["WebSocketApi"]} --stage-name ${var.websocket_api_stage} --description 'Deployed by ${var.lifeomic_project}'"
  }
}
```

**Comment by rpstreef** [comment]
> 
> 
> welcome to refer to a [websocket template](https://github.com/systemaccounting/mxfactorial/blob/0e6237fd53fa072876e8d08d8ae868bed910c2bf/infrastructure/cloudformation/websockets.yaml) in cloudformation until merged
> 
> [cli documentation](https://github.com/systemaccounting/mxfactorial/blob/0e6237fd53fa072876e8d08d8ae868bed910c2bf/infrastructure/cloudformation/README.md) available

I'm aware of the alternatives, that's not the point The point here is that we need this properly integrated and that we're close to 1 year since release and it's still not there. 
That should concern anyone using Terraform, it means I may have to start using alternatives all together because the updates are too infrequent.

**Comment by mbonig** [comment]
> > welcome to refer to a [websocket template](https://github.com/systemaccounting/mxfactorial/blob/0e6237fd53fa072876e8d08d8ae868bed910c2bf/infrastructure/cloudformation/websockets.yaml) in cloudformation until merged
> > [cli documentation](https://github.com/systemaccounting/mxfactorial/blob/0e6237fd53fa072876e8d08d8ae868bed910c2bf/infrastructure/cloudformation/README.md) available
> 
> I'm aware of the alternatives, that's not the point The point here is that we need this properly integrated and that we're close to 1 year since release and it's still not there.
> That should concern anyone using Terraform, it means I may have to start using alternatives all together because the updates are too infrequent.

Already quit Terraform for this exact reason. Having to split my IaC across TF and CFN to get what I need is dumb. Just going to use CFN (with the CDK). I recommend anyone else do the same until the teams overseeing the AWS Provider can get caught up.

**Comment by krotscheck** [comment]
Ditto - though I went to ansible, it's far easier to extend.

**Comment by davidhoustcdk** [comment]
339 days in.. is there any activity on this?

**Comment by marcoreni** [comment]
After the release of HTTP APIs I think it would be useful to push this forward.

I'm willing to contribute on this, but I see there are already PR opened for multiple resources.

@ewbankkit what's the status of those?

**Comment by ewbankkit** [comment]
@marcoreni All the accceptance tests pass for the WebSocket APIs (last time I ran them).
After a brief initial look at the changes for HTTP APIs it doesn't look like we'll need to have additional resources over the ones I implemented in the various PRs but a number of them will have to be augmented and additional acceptance tests added. We should also think about how to handle [API migration/import](https://docs.aws.amazon.com/apigateway/latest/developerguide/http-api-import.html).

**Comment by marcoreni** [comment]
I think that REST/HTTP migration is not responsibility of TF since according to  in AWS you must export the apigatewayv1 JSON and then import it to apigatewayv2.

AFAICS `apigateway` does *not* have an SDK equivalent to the `import-api` cli command, so even if we could create a datasource/extend the existing `api_gateway_rest_api` datasource to expose the openapi3 json using [APIGateway.GetExport](https://docs.aws.amazon.com/sdk-for-go/api/service/apigateway/#APIGateway.GetExport), there would be no way to import it in `apigatewayv2`.

As per the new implementations, we can fork from the PR while waiting for them to be merged (maybe we can ping maintainers on those?)

**Comment by bionicles** [comment]
Hi, I would definitely use both HTTP and WS options immediately for a graph db project ... would it be possible to merge these soon?

**Comment by rpstreef** [comment]
When will this be released since the acceptance tests have passed?

**Comment by finiteinfinity** [comment]
Would be nice to get some sort of update regarding when this will be released

**Comment by vcalmic** [comment]
Are there any updates regarding when this feature will be release?

**Comment by eistrati** [comment]
This feature would be really useful for my team. Any updates when it will be published?

**Comment by rpj** [comment]
FWIW, this is the pertinent `awscli` invocations  of a script I've written to workaround this issue, in the event it might help others who need to use API Gateway WebSockets now (I can't share the entire thing but hopefully it's enough to be helpful):

```
aws apigatewayv2 --profile $PROFILE --region $REGION \
  create-api --name $API_NAME --protocol-type WEBSOCKET \
  --route-selection-expression '$request.body.channel' > $TMP

API_ID=$(cat $TMP | jq -r '.ApiId')
API_EP=$(cat $TMP | jq -r '.ApiEndpoint')

aws apigatewayv2 --profile $PROFILE --region $REGION \
  create-integration --api-id $API_ID \
  --credentials-arn $GW_CREDS_ARN \
  --integration-type AWS_PROXY --integration-method POST \
  --integration-uri $INVOKE_ARN > $TMP

INT_ID=$(cat $TMP | jq -r '.IntegrationId')

aws apigatewayv2 --profile $PROFILE --region $REGION \
  create-route --api-id $API_ID --route-key '$connect' \
  --authorization-type 'NONE' --no-api-key-required \
  --target "integrations/$INT_ID" > $TMP

aws apigatewayv2 --profile $PROFILE --region $REGION \
  create-route --api-id $API_ID --route-key '$disconnect' \
  --target "integrations/$INT_ID" > $TMP

aws apigatewayv2 --profile $PROFILE --region $REGION \
  create-route --api-id $API_ID --route-key '$default' \
  --target "integrations/$INT_ID" > $TMP

DEF_ROUTE_ID=$(cat $TMP | jq -r '.RouteId')

aws apigatewayv2 --profile $PROFILE --region $REGION \
  create-deployment --api-id $API_ID > $TMP

DEPLOYMENT_ID=$(cat $TMP | jq -r '.DeploymentId')

aws apigatewayv2 --profile $PROFILE --region $REGION \
  create-stage --api-id $API_ID --deployment-id $DEPLOYMENT_ID \
  --default-route-settings "DataTraceEnabled=true,DetailedMetricsEnabled=true,LoggingLevel=$WS_GW_LOGGING_LEVEL" \
  --stage-name $STAGE_NAME > $TMP
```

**Comment by deding** [comment]
My shop is all-in on terraform and our dev teams are wanting to start work on a websockets api.  We really don't want to have to manage this one resource manually while everything else is in terraform.  Any updates on when this will be released?

**Comment by mbonig** [comment]
Until it's merged you could compile this PR into a custom provider and use that ... not a great long-term solution but could allow you to get started in a small part.

**Comment by deding** [comment]
Thanks @mbonig for the suggestion - might just have to do that so our team doesn’t accumulate even more tech debt.  I hope this gets released soon

**Comment by mvisonneau** [comment]
@mbonig or @ewbankkit, is there a branch on your forks you could recommend in order to compile it with most of the new resources?

**Comment by ewbankkit** [comment]
@mvisonneau The working example https://github.com/terraform-providers/terraform-provider-aws/pull/9533 uses the branch `ewbankkit/issue-7004.aws_api_gateway_v2_stage` from https://github.com/terraform-providers/terraform-provider-aws/pull/9232.

**Comment by rpstreef** [comment]
So why is this not getting integrated yet?

Generally I'm wondering, the community voting system doesn't seem to work at advertised. This issue and several others have been waiting for a code release for months.
I think it would help everyone to give some clarity in terms of priorities and what is getting worked on. Thanks

**Comment by mvisonneau** [comment]
cross referencing here that #8842 has now been merged and released as part of [v2.53.0](https://github.com/terraform-providers/terraform-provider-aws/releases/tag/v2.53.0)

many thanks @mbonig @ewbankkit & @gdavison! 🙇 

hopefully the remaining resources will be able to get reviewed and integrated soon too! 🙏

**Comment by TheRealChrisThomas** [comment]
Trying to get an update on this issue? I see that the base `aws_apigatewayv2_api` has been created, but I'm trying to use the HTTP version so need other pieces referenced above. Thanks for all the work so far on this!

**Comment by codingbadger** [comment]
Is there any indication as to when this will be released? HTTP API Gateways were made General Release on 12th March 2020.

**Comment by mbonig** [comment]
> Is there any indication as to when this will be released? HTTP API Gateways were made General Release on 12th March 2020.

Keep in mind this is totally separate than the HTTP API Gateway that was just released... this is for V2 of API Gateway, the Websockets version released back in 2018...

**Comment by codingbadger** [comment]
@mbonig Oh my mistake, the AWS API for API Gateway v2 (https://docs.aws.amazon.com/apigateway/latest/developerguide/api-ref.html) is for both WebSocket and HTTP so I incorrectly assumed that would do the same.  

If this only handles WebSocket where is the HTTP implementation? Or does that not exist yet>

**Comment by mbonig** [comment]
@codingbadger Well my bad too.. didn't realize they had put the HTTP API in the V2 endpoints. This is just for the Websocket portions and I doubt it'll have anything specific for the HTTP API.

**Comment by martincastrocm** [comment]
Hi there! I want to use `aws_apigatewayv2_api` but need all the complementary components too. Is there any estimative release date of these resources?

**Comment by maryelizbeth** [comment]
Hi Y'all, AWS Product Manager here!

As we've now merged the `_api` bit of `apigatewayv2`, the team will be working with the authors to review and merge the remaining collection of PRs for the Websocket API over the next month. 

We'll then start evaluating the HTTP API ([#11148](https://github.com/terraform-providers/terraform-provider-aws/issues/11148)), but as of today we can't provide a timeline for release. 

Hope that clears up the confusion around the timing of this feature. Feel free to @ me for any questions regarding the timing or prioritization. I'll update this issue if things change regarding this feature's release.

**Comment by ebarault** [comment]
@ewbankkit thanks of all for pushing all this through!

do you happen to have a single branch where all the feature branches regarding api_gateway_v2 are all merged in ?
I'm currently compiling the aws provider from your different branches and it will be easier to test by building only one provider gathering all the new features.

**Comment by ebarault** [comment]
Splendid !
3 new ressources are available as of today:
**aws_apigatewayv2_authorizer
aws_apigatewayv2_integration
aws_apigatewayv2_model**

**Comment by ebarault** [comment]
hi @ewbankkit 
the module **aws_apigatewayv2_api** creates 3 routes by default (or they get created automatically by aws)
`$default`
`$connect`
`$disconnect`
but the module does not export in its outputs the RouteId key for each of those routes, and we don't have a data source to retrieve them

Hence, when the module aws_apigatewayv2_route will be released, we won't be able to update the default routes with their target integrations.

How would you see the next step here? 
- A data source to get the routes by name?
- Adding the default routes ids as outputs of the api module ?

**Comment by ewbankkit** [comment]
@ebarault When a WebSocket API is created via Terraform those 3 routes aren't created implicitly:

```console
$ terraform apply

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  + aws_apigatewayv2_api.test
      id:                           <computed>
      api_endpoint:                 <computed>
      api_key_selection_expression: "$request.header.x-api-key"
      arn:                          <computed>
      execution_arn:                <computed>
      name:                         "example-ws-api"
      protocol_type:                "WEBSOCKET"
      route_selection_expression:   "$request.body.action"


Plan: 1 to add, 0 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

aws_apigatewayv2_api.test: Creating...
  api_endpoint:                 "" => "<computed>"
  api_key_selection_expression: "" => "$request.header.x-api-key"
  arn:                          "" => "<computed>"
  execution_arn:                "" => "<computed>"
  name:                         "" => "example-ws-api"
  protocol_type:                "" => "WEBSOCKET"
  route_selection_expression:   "" => "$request.body.action"
aws_apigatewayv2_api.test: Creation complete after 1s (ID: p5axu0jzmb)

Apply complete! Resources: 1 added, 0 changed, 0 destroyed.
kit@ub18043x64-dev:~/tmp/apigatewayv2$ aws --region us-west-2 apigatewayv2 get-routes --api-id p5axu0jzmb
{
    "Items": []
}
```

but they do show up in the AWS console (albeit subtly grayed out) and no doubt there will be many users who click on those routes, which does cause their implicit creation (and also additional integrations etc.). So, yes we should add data sources for all corresponding the API Gateway v2 resources - Phase 2 😄.

**Comment by Ludonope** [comment]
I find the name `credentials_arn` confusing in `aws_apigatewayv2_integration`.
Shouldn't it be `execution_role` instead?

**Comment by ebarault** [comment]
@ewbankkit ohhhh that's twisted! i just destroyed my apigw_v2, recreated it, and a call to `aws apigatewayv2 get-routes --api-id xxxxxxxxx` confirms that there is no route created by default.

As soon as i click on one of those + signs, a route gets created:
<img width="205" alt="image" src="https://user-images.githubusercontent.com/12000935/79110222-41b3b580-7d7a-11ea-9a6f-61ff465800da.png">

**Comment by ebarault** [comment]
@Ludonope the name comes from aws:
https://docs.aws.amazon.com/cli/latest/reference/apigatewayv2/create-integration.html

**Comment by ewbankkit** [comment]
@Ludonope As @ebarault mentions, we aim to maintain fidelity with the underlying AWS API. If you feel the documentation can be clearer, please open a linked issue. Thanks.

**Comment by christhomas** [comment]
When I create a new "aws_apigatewayv2_api" and "aws_apigatewayv2_domain_name" I'm not given any hosted zone or domain name information like I do with version 1. 

How am I supposed to connect this api to a route53 record and use it? There doesn't appear to be any information in the documentation about how I could do this either.

**Comment by ewbankkit** [comment]
@christhomas Thanks for raising this - Route 53 hosted zone and API Gateway domain name are available from the underlying AWS [`GetDomainName`](https://docs.aws.amazon.com/apigatewayv2/latest/api-reference/domainnames-domainname.html#GetDomainName) API.
Could you please open a separate issue for this and we should be able to address promptly.

**Comment by ebarault** [comment]
cross-referencing the aforementioned issue: https://github.com/terraform-providers/terraform-provider-aws/issues/12892

**Comment by christhomas** [comment]
I also think this issue should be cross-referenced too: https://github.com/terraform-providers/terraform-provider-aws/issues/12894

**Comment by ewbankkit** [comment]
https://github.com/terraform-providers/terraform-provider-aws/pull/9461 was the final PR for new resources. I think we'll need data sources corresponding to some of the API Gateway v2 resources, but @bflad, @maryelizbeth do you want to close this issue now and open separate issues for additional work?

**Comment by gdavison** [comment]
Hello everyone,

All of the functionality linked to this Issue has either been released, or will be released as part of v2.59.0 later this week. Namely, the resource types

* aws_apigatewayv2_api
* aws_apigatewayv2_api_mapping
* aws_apigatewayv2_authorizer
* aws_apigatewayv2_deployment
* aws_apigatewayv2_domain_name
* aws_apigatewayv2_integration
* aws_apigatewayv2_integration_response
* aws_apigatewayv2_model
* aws_apigatewayv2_route
* aws_apigatewayv2_route_response
* aws_apigatewayv2_stage

which support ApiGateway v2 WebSocket APIs. Follow #11148 for the enhancements needed for HTTP APIs.

For further feature requests, documentation updates, or bug reports with this functionality, please create a [new GitHub issue](https://github.com/terraform-providers/terraform-provider-aws/issues/new/choose) following the template for triage. Thanks!

**Comment by jacktkeys** [comment]
I have had a couple of observations the last couple of days trying to implement API Gateway V2 in Terraform.
1. aws_apigatewayv2_api is missing the CORS configuration block
2. There is no implementation of a VPC link for v2.

I created a functioniong API gateway in the console pointing to an internal ALB. I went to list all of the resources in the AWS cli. Everything was there except for the VPC link. I upgraded AWS cli to verison 1.18.41 and now the apigatewayv2 subcommand has references to vpc link.

**Comment by ewbankkit** [comment]
@jacktkeys `aws_apigatewayv2_api` has been implemented in https://github.com/terraform-providers/terraform-provider-aws/pull/12452 and new resource `aws_apigatewayv2_vpc_link` has been implemented in https://github.com/terraform-providers/terraform-provider-aws/pull/12577.
Both PRs have been merged and will be released with provider v2.59.0.

**Comment by jacktkeys** [comment]
@ewbankkit Thanks for pointing me to those. I look forward to the next release then.

**Comment by christhomas** [comment]
I really really really want api_mapping right now, it's what is causing my final problem on this ticket: https://github.com/terraform-providers/terraform-provider-aws/issues/12892

**Comment by christhomas** [comment]
There is an issue with deployments that I've found, it's a bit tricky to explain, I've created a ticket to try to get what I think is happening across: https://github.com/terraform-providers/terraform-provider-aws/issues/12961
